src/rust/engine/options/src/lib.rs

```diff
@@ -158,27 +159,62 @@ pub struct OptionParser {
 
 impl OptionParser {
     pub fn new(env: Env, args: Args) -> Result<OptionParser, String> {
+        let buildroot = BuildRoot::find()?;
+        let buildroot_string = String::from_utf8(buildroot.as_os_str().as_bytes().to_vec())
+            .map_err(|e| {
+                format!(
+                    "Failed to decode build root path {}: {}",
+                    buildroot.display(),
+                    e
+                )
+            })?;
+
+        let mut seed_values = HashMap::from([
+            ("buildroot".to_string(), buildroot_string.clone()),
+            ("homedir".to_string(), shellexpand::tilde("~").into_owned()),
+            ("user".to_string(), whoami::username()),
+        ]);
+        seed_values.extend(
+            env.env
+                .iter()
+                .map(|(k, v)| (format!("env.{k}", k = k), v.clone())),
+        );
+
         let mut sources: BTreeMap<Source, Rc<dyn OptionsSource>> = BTreeMap::new();
         sources.insert(Source::Env, Rc::new(env));
         sources.insert(Source::Flag, Rc::new(args));
         let mut parser = OptionParser {
             sources: sources.clone(),
         };
 
-        let config_path = BuildRoot::find()?.join("pants.toml");
+        fn path_join(a: &str, b: &str) -> String {
+            format!("{}{}{}", a, path::MAIN_SEPARATOR, b)
+        }
+
+        let default_config_path = path_join(&buildroot_string, "pants.toml");
         let repo_config_files = parser.parse_string_list(
             &option_id!("pants", "config", "files"),
-            &[
-                std::str::from_utf8(config_path.as_os_str().as_bytes()).map_err(|e| {
-                    format!(
-                        "Failed to decode build root path {}: {}",
-                        config_path.display(),
-                        e
-                    )
-                })?,
-            ],
+            &[&default_config_path],
         )?;
-        let mut config = Config::merged(&repo_config_files)?;
+
+        let default_workdir = path_join(&buildroot_string, ".pants.d");
+        let default_distdir = path_join(&buildroot_string, "dist");
+        seed_values.insert(
+            "pants_workdir".to_string(),
+            parser
+                .parse_string(&option_id!("pants", "workdir"), &default_workdir)?
+                .value
+                .clone(),
+        );
+        seed_values.insert(
+            "pants_distdir".to_string(),
+            parser
+                .parse_string(&option_id!("pants", "distdir"), &default_distdir)?
+                .value
+                .clone(),
+        );
```

Comment:
Minor syntactic thing, but could the code be shuffled around so that these could be included directly in the `HashMap::from` call, instead of being dedicated inserts?

Reply:
No: `seed_values` need to be extended by the env vars before we create the initial config, because it consumes env, but these values can only be computed after we create the initial config.  We could initialize seed_values to empty, add the env vars, and then insert the other seed values here, but we can't add them in a `HashMap::from`.

Reply:
OK, done that, so at least all the non-env seeds are seeded in one place.